design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merges
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
+	// +required
+	ServiceRef SupportServiceReference `json:"serviceRef"`
+
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+
+	// Timeout configures maximum time to wait for a check response from the authorization server.
+	// Timeout durations are expressed in the Go [Duration format](https://godoc.org/time#ParseDuration).
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// The string "infinity" is also a valid input and specifies no timeout.
+	//
+	// +optional
+	Timeout string `json:"timeout,omitempty"`
+}
+
+// VirtualHost appears at most once. If it is present, the object is considered
+// to be a "root".
+type VirtualHost struct {
+        ...
+
+	// This field configures a support service to perform
+	// authorization for this virtual host. Authorization can
+	// only be configured on virtual hosts that have TLS enabled.
+	// If the TLS configuration requires client certificate
+	///validation, the client certificate is always included in the
+	// authentication check request.
+	//
+	// +optional
+	Authorization *AuthorizationServer `json:"authorization,omitempty"`
+}
+
+// Route contains the set of routes for a virtual host.
+type Route struct {
+        ...
+        // AuthPolicy updates the authorization policy for client
+	// requests that match this route.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+}
+```
+
+An operator configures authorization on a root `HTTPProxy` by setting the `AuthorizationServer` field.
+Setting this field without also setting the `TLS` field is an error.
+
+The `AuthorizationServer` field carries and optional default authorization policy.
+This policy sets the default policy for each subsequent `Route`.
+If `authPolicy.Disabled` is true, included `Routes` will not require authorization unless they configure their own authorization policy with `Disabled` set to false.
```

Comment:
How does `permitInsecure` play into this?

Reply:
That's a good point, I'll check on that.

Reply:
If a route can be insecure, then I don't see a way for the upstream service to guarantee that it can know whether a request has been authorized or not. I updated to specify that insecure routes must disable authorization.